keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades/README.md

```diff
@@ -0,0 +1,860 @@
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+In addition to using `pod-template-hash` added by the Deployment controller, 
+users can also provide the customized key in  `MatchLabelKeys` to identify 
+which pods should be grouped. If so, the user needs to ensure that it is 
+correct and not duplicated with other unrelated workloads.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+A new field named `MatchLabelKeys` will be added to `TopologySpreadConstraint`. 
+Currently, when scheduling a pod, the `LabelSelector` defined in the pod is used 
+to identify the group of pods over which spreading will be calculated. 
+`MatchLabelKeys` adds another constraint to how this group of pods is identified: 
+the scheduler will use those keys to look up label values from the incoming pod; 
+and those key-value labels are ANDed with `LabelSelector` to select the group of 
+existing pods over which spreading will be calculated.
+
+A new field named `MatchLabelKeys` will be introduced to`TopologySpreadConstraint`:
+```go
+type TopologySpreadConstraint struct {
+	MaxSkew           int32
+	TopologyKey       string
+	WhenUnsatisfiable UnsatisfiableConstraintAction
+	LabelSelector     *metav1.LabelSelector
+
+	// MatchLabelKeys is a set of pod label keys to select the pods over which 
+	// spreading will be calculated. The keys are used to lookup values from the
+	// incoming pod labels, those key-value labels are ANDed with `LabelSelector`
+	// to select the group of existing pods over which spreading will be calculated
+	// for the incoming pod. Keys that don't exist in the incoming pod labels will
+	// be ignored.
+	MatchLabelKeys []string
+}
+```
+
+Examples of use are as follows:
+```yaml
+     topologySpreadConstraints:
+        - maxSkew: 1
+          topologyKey: kubernetes.io/hostname
+          whenUnsatisfiable: DoNotSchedule
+          matchLabelKeys:
+            - app
+            - pod-template-hash
+```
+
+The scheduler plugin `PodTopologySpread` will obtain the labels from the pod 
+labels by the keys in `matchLabelKeys`. The obtained labels will be merged 
+to `labelSelector` of `topologySpreadConstraints` to filter and group pods. 
+The pods belonging to the same group will be part of the spreading in 
+`PodTopologySpread`.
+
+Finally, the feature will be guarded by a new feature flag. If the feature is 
+disabled, the field `matchLabelKeys` is preserved if it was already set in the 
+persisted Pod object, otherwise it is silently dropped; moreover, kube-scheduler 
+will ignore the field and continue to behave as before.
+
+### Test Plan
```

Comment:
Please adopt to the new test plan guidelines:

https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#test-plan